src/anim/controller/anim-blend-tree.js

```diff
@@ -204,6 +94,176 @@ class AnimBlendTree extends AnimNode {
+ * @description Create a new BlendTree1D.
+ */
+class AnimBlendTreeDirectional2D extends AnimBlendTree {
+    pointAngleCache(i, j) {
+        if (!this._pointCache) this._pointCache = {};
+        var pointKey = `${i}${j}`;
+        if (!this._pointCache[pointKey]) {
+            this._pointCache[pointKey] = AnimBlendTree.getAngleRad(this._children[i].point, this._children[j].point);
+        }
+        return this._pointCache[pointKey];
+    }
+
+    pipjCache(i, j) {
+        if (!this._pipjCache) this._pipjCache = {};
+        var pointKey = `${i}${j}`;
+        if (!this._pointCache[pointKey]) {
+            this._pointCache[pointKey] = new Vec2(
+                (this._children[j].pointLength - this._children[i].pointLength) / ((this._children[j].pointLength + this._children[i].pointLength) / 2),
+                AnimBlendTree.getAngleRad(this._children[i].point, this._children[j].point) * 2.0
+            );
+        }
+        return this._pointCache[pointKey];
+    }
+
+    calculateWeights() {
+        var i, j, pi, pipj, parameterValues, minj, result, weightSum;
+        parameterValues = this.currentParameterValues();
+        if (this.parametersEqual(parameterValues)) return;
+        this._parameterValues = parameterValues;
+        this._p.set(this._parameterValues);
+        var pLength = p.length();
+        weightSum = 0.0;
+        for (i = 0; i < this._children.length; i++) {
+            pi = this._children[i].point;
+            var piLength = this._children[i].pointLength;
+            minj = Number.MAX_VALUE;
+            for (j = 0; j < this._children.length; j++) {
+                if (i === j) continue;
+                var pjLength = this._children[j].pointLength;
+                pipj = this.pipjCache(i, j);
+                this._pip.set((pLength - piLength) / ((pjLength + piLength) / 2), AnimBlendTree.getAngleRad(pi, this._p) * 2.0);
+                result = AnimBlendTree.clamp(1.0 - Math.abs((this._pip.dot(pipj) / pipj.lengthSq())), 0.0, 1.0);
+                if (result < minj) minj = result;
+            }
+            this._children[i].weight = minj;
+            weightSum += minj;
+        }
+        for (i = 0; i < this._children.length; i++) {
+            this._children[i].weight = this._children[i]._weight / weightSum;
+        }
+    }
+}
+
+
+/**
+ * @private
+ * @class
+ * @name AnimBlendTreeDirect
+ * @classdesc An AnimBlendTree that calculates normalised weight values based on the total weight
+ * @description Create a new BlendTree1D.
+ */
+class AnimBlendTreeDirect extends AnimBlendTree {
+    calculateWeights() {
+        var i, parameterValues, weightSum;
+        parameterValues = this.currentParameterValues();
+        if (this.parametersEqual(parameterValues)) return;
+        this._parameterValues = parameterValues;
+        weightSum = 0.0;
+        for (i = 0; i < this._children.length; i++) {
+            weightSum += AnimBlendTree.clamp(this._parameterValues[i], 0.0, Number.MAX_VALUE);
```

Comment:
instead of clamp, just use max(value, 0) - clamping high range to MAX_VALUE is not useful

Reply:
Done 👍